ios-base/Networking/APIService.swift

```diff
@@ -0,0 +1,180 @@
+                       onFailure: ((Error, Response?) -> Void)? = nil) where T: Codable {
+    provider.request(target, completion: { [weak self] result in
+      guard let self = self else { return }
+
+      switch result {
+      case let .success(moyaResponse):
+        self.handleSuccess(with: moyaResponse, at: keyPath, onSuccess, onFailure)
+      case let .failure(error):
+        self.handleError(with: error, onFailure)
+      }
+    })
+  }
+
+  /**
+   Makes a request to the provided target
+   */
+  open func request(for target: ProviderType,
+                    onSuccess: @escaping (_ response: Response) -> Void,
+                    onFailure: ((Error, Response?) -> Void)? = nil) {
+    provider.request(target, completion: { [weak self] result in
+      guard let self = self else { return }
+
+      switch result {
+      case let .success(moyaResponse):
+        onSuccess(moyaResponse)
+      case let .failure(error):
+        self.handleError(with: error, onFailure)
+      }
+    })
+  }
+
+  private func handleSuccess<T>(with response: Response,
+                                at keyPath: String? = nil,
+                                _ onSuccess: @escaping (_ result: T, _ response: Response) -> Void,
+                                _ onFailure: ((Error, Response?) -> Void)? = nil) where T: Decodable {
+    do {
+      let filteredResponse = try response.filterSuccessfulStatusCodes()
+      let decodedResult = try filteredResponse.map(T.self,
+                                                   atKeyPath: keyPath,
+                                                   using: jsonDecoder,
+                                                   failsOnEmptyData: true)
+
+      onSuccess(decodedResult, filteredResponse)
+    } catch (let error) {
+      handleError(with: error, onFailure)
+    }
+  }
+
+  private func handleError(with error: Error, _ onFailure: ((Error, Response?) -> Void)? = nil) {
+    guard let moyaError = error as? MoyaError else {
+      onFailure?(error, nil)
+      return
+    }
+    switch moyaError {
+    case .imageMapping(let response):
+      onFailure?(error, response)
+    case .jsonMapping(let response):
+      onFailure?(error, response)
+    case .statusCode(let response):
+      if response.statusCode == StatusCode.unauthorized {
+        AppDelegate.shared.unexpectedLogout()
+      }
+      let decodedError = APIError.from(response: response)
+      onFailure?(decodedError ?? error, response)
+    case .stringMapping(let response):
+      onFailure?(error, response)
+    case .objectMapping(let error, let response):
+      onFailure?(error, response)
+    case .encodableMapping(let error):
+      onFailure?(error, nil)
+    case .underlying(let error, let response):
+      onFailure?(error, response)
+    case .requestMapping:
+      onFailure?(error, nil)
+    case .parameterEncoding(let error):
+      onFailure?(error, nil)
```

Comment:
lets group the cases that execute the same code for readability and avoiding code repetition